packages/truffle-debugger/lib/data/reducers.js

```diff
@@ -187,30 +196,104 @@ function mappingKeys(state = DEFAULT_MAPPING_KEYS, action) {
+      baseType = Definition.restorePtr(baseType);
+
+      debug("slot %o", slot);
+      let hexSlotAddress = Conversion.toHexString(
+        slotAddress(slot),
+        EVM.WORD_SIZE
+      );
+      let parentAddress = slot.path
+        ? Conversion.toHexString(slotAddress(slot.path), EVM.WORD_SIZE)
+        : undefined;
 
-      return {
+      //this is going to be messy and procedural, sorry.  but let's start with
+      //the easy stuff: create the new address if needed, clone if not
+      let newState = {
         decodingStarted: state.decodingStarted,
-        byId: {
-          ...state.byId,
+        byAddress: {
+          ...state.byAddress,
+          [address]: {
+            byType: {
+              ...(state.byAddress[address] || { byType: {} }).byType
+            }
+          }
+        }
+      };
 
-          // add new key to set of keys already defined
-          [id]: [
-            ...new Set([
-              //set for uniqueness
-              ...(state.byId[id] || []),
-              key
-            ])
-          ]
+      //now, let's add in the new type, if needed
+      newState.byAddress[address].byType = {
+        ...newState.byAddress[address].byType,
+        [typeIdentifier]: {
+          bySlotAddress: {
+            ...(
+              newState.byAddress[address].byType[typeIdentifier] || {
+                bySlotAddress: {}
+              }
+            ).bySlotAddress
+          }
         }
       };
 
+      let oldSlot =
+        newState.byAddress[address].byType[typeIdentifier].bySlotAddress[
+          hexSlotAddress
+        ];
+      //yes, this looks strange, but we haven't changed it yet except to
+      //clone or create empty (and we don't want undefined!)
+      //now: is there something already there or no?  if no, we must add
+      if (oldSlot === undefined) {
+        let newSlot;
+        debug("parentAddress %o", parentAddress);
+        if (
+          parentAddress !== undefined &&
+          newState.byAddress[address].byType[baseType] &&
+          newState.byAddress[address].byType[baseType].bySlotAddress[
+            parentAddress
+          ]
+        ) {
+          //if the parent is already present, use that instead of the given
+          //parent!
+          newSlot = {
+            ...slot,
+            path:
+              newState.byAddress[address].byType[baseType].bySlotAddress[
+                parentAddress
+              ]
+          };
+        } else {
+          newSlot = slot;
+        }
+        newState.byAddress[address].byType[typeIdentifier].bySlotAddress[
+          hexSlotAddress
+        ] = newSlot;
+      }
+      //if there's already something there, we don't need to do anything
+
+      return newState;
+
     case actions.RESET:
-      return DEFAULT_MAPPING_KEYS;
+      return DEFAULT_PATHS;
+
+    case actions.LEARN_ADDRESS:
+      return {
+        decodingStarted: state.decodingStarted,
```

Comment:
```suggestion
        ...state,
```

Reply:
Done

Reply:
Also made the analogous change elsewhere in that reducer